### ISSUE TITLE
Red Mushrooms does not clear from selected slot when dying from them; possible item duplication glitch

### ISSUE THREAD

**Issue by juozaspo** [minetest/minetest#6546]
Using red mushrooms (eating) does not clear all the current slot. The current item (red mushrooms that left) is still present in death screen. Collecting multiple bones leads to red mushroom duplication.
Screenshots:
![screenshot_20171020_183253](https://user-images.githubusercontent.com/6626242/31830849-28d525fa-b5ca-11e7-92f4-33e9531096b8.png)
![screenshot_20171020_183258](https://user-images.githubusercontent.com/6626242/31830850-28fd5b6a-b5ca-11e7-9fef-f7b1015e6d9f.png)
![screenshot_20171020_183306](https://user-images.githubusercontent.com/6626242/31830851-2922e470-b5ca-11e7-8400-4cae6f6ffa5e.png)
![screenshot_20171020_185646](https://user-images.githubusercontent.com/6626242/31830852-294fb036-b5ca-11e7-899f-c63733420358.png)
![screenshot_20171020_185852](https://user-images.githubusercontent.com/6626242/31830853-2976090c-b5ca-11e7-8d90-e0481330b925.png)

**Comment by juozaspo** [comment]
It appears the bug is known in minetest_game bug tracker, known bug; closing this then.

**Comment by jastevenson303** [comment]
That's true, but @rubenwardy thought it (https://github.com/minetest/minetest_game/issues/1900) was an engine issue, and so should likely be tracked here in addition, or instead of.

**Comment by juozaspo** [comment]
@jastevenson303 good idea, I'll reopen it then

**Comment by rubenwardy** [comment]
Is the selected item in a chest? If so, this should be fixed in 0.5.0-dev due to distance limitations on inventory action

**Comment by JustinLaw64** [comment]
It looks like the issue is caused by the way the area surrounding this line of code is ordered.

https://github.com/minetest/minetest/blob/5560ec8178053b27adac6f277f539975d46267b8/builtin/game/item.lua#L470

Here's what I think goes on when a player has a health of 5 or less and they eat another mushroom:

1. minetest.do_item_eat or core.do_item_eat is called for the mushroom with the stack
2. set_hp is called at that line, setting their health at or below 0, killing the player and triggering events such as the ones in the bones mod.
3. The bones mod clears everything from the player's inventory from further down the call stack from set_hp.
4. The function returns the item stack with an item removed, causing the Minetest engine to overwrite the now empty slot.

I'm trying to figure out how to kill the player after the slot is written to. Not yet sure how to do this though.

**Comment by tenplus1** [comment]
@JustinLaw64 - We could split the coe.do_item_eat() function into 2 parts, first handles inventory stack and 2nd will deal damage afterwards.

**Comment by tenplus1** [comment]
@JustinLaw64 - The following change to that line seems to work:
```
minetest.after(0.1, function()
	user:set_hp(user:get_hp() + hp_change)
end, user, hp_change)

**Comment by HybridDog** [comment]
Maybe the problem needs to be fixed in the bones mod so that the inventory is put into a bones node only when other (relevant) lua functions have finished executing. 

1. Some function gets the inventory in its current state
2. This function kills the player
3. Minetest executes functions registered via `minetest.register_on_dieplayer`, e.g. the bones spawning.
4. The function issues an inventory change based on the inventory state it found previously

`minetest.after` may be used for this purpose, but I don't know if there are negative side effects if the inventory is moved after the player has respawned. 

1. Some function gets the inventory in its current state
2. This function kills the player
3. Bones remembers the position where to create the bones node and invokes a minetest.after(0, […])
4. The function issues an inventory change based on the inventory state it found previously
5. (Hopefully) unrelated functions are executed
6. Minetest executes the function which bones passed to minetest.after, which is moving the inventory to a bones node.

It may be required to use `minetest.register_on_respawnplayer` in addition to `minetest.after` so that the player never has his/her inventory after respawning.

**Comment by SmallJoker** [comment]
Found the culprit by accident.
https://github.com/minetest/minetest/blob/5c89a0e12a1e679180b14bf92bdcdb1614e3982e/src/network/serverpackethandler.cpp#L1316-L1320

`setWieldedItem` could be skipped when the player was killed by `on_use`, but then what should be done about the ItemStack change? Proposal: only send `SendPlayerHPOrDie` each server step, just like inventory updates. The player will be dead for at least one server tick anyway (due to formspec).

**Comment by sfan5** [comment]
SendInventory isn't the problem here, or are you suggesting to throw away the returned new item when the player dies before the callback finishes?

**Comment by Wuzzy2** [comment]
Yes, I was suggesting that. To be precise, disallow/ignore all attempted inventory changes while the player is dead. 

But I'm not sure if this woud actually work in practice. It's a quick ad-hoc solution that just came to mind.
